package/cell_modem_daemon/src/at_command_utils.c

```diff
@@ -0,0 +1,336 @@
+const char *at_serial_port_command_result(at_serial_port_command_t *at_command)
+{
+  return at_command->result;
+}
+
+/* The following from Telit HE910 AT Command Reference
+ * +CSQ: <rssi>,<ber>
+ * where
+ * <rssi> - received signal strength indication
+ *  0 - (-113) dBm or less
+ *  1 - (-111) dBm
+ *  2..30 - (-109)dBm..(-53)dBm / 2 dBm per step
+ *  31 - (-51)dBm or greater
+ *  99 - not known or not detectable
+ * <ber> - bit error rate (in percent)
+ *  0 - less than 0.2%
+ *  1 - 0.2% to 0.4%
+ *  2 - 0.4% to 0.8%
+ *  3 - 0.8% to 1.6%
+ *  4 - 1.6% to 3.2%
+ *  5 - 3.2% to 6.4%
+ *  6 - 6.4% to 12.8%
+ *  7 - more than 12.8%
+ *  99 - not known or not detectable
+ */
+static const s8 signal_to_dbm_lookup[] = {
+  -113, // zero case
+  -111, -109, -107, -105, -103, -101, -99, -97, -95, -93,  // 1 - 30 cases
+  -91,  -89,  -87,  -85,  -83,  -81,  -79,  -77, -75, -73,
+  -71, -69,  -67,  -65,  -63,  -61,  -59,  -57,  -55, -53,
+  -51 // 31 case
+};
+
+static s8 modem_signal_strength_to_dbm(u32 signal_raw)
+{
+  s8 signal_dbm = 0; // unknown will remain 0
+
+  if (signal_raw < sizeof(signal_to_dbm_lookup) / sizeof(signal_to_dbm_lookup[0])) {
+    signal_dbm = signal_to_dbm_lookup[signal_raw];
+  }
+  return signal_dbm;
+}
+
+// Splits the difference from above
+static const float error_to_percent_lookup[] = {
+  0.2f, 0.3f, 0.6f, 1.2f, 2.4f, 4.8f, 9.6f, 12.8f
+};
+
+static float modem_error_rate_to_percent(u32 error_raw)
+{
+  float error_percent = 0.0; // unknown will remain 0
+
+  if (error_raw < sizeof(error_to_percent_lookup) / sizeof(error_to_percent_lookup[0])) {
+    error_percent = error_to_percent_lookup[error_raw];
+  }
+  return error_percent;
+}
+
+// This block should be refactored
+int at_command_report_signal_quality(at_serial_port_t *port,
+                                     s8 *signal_strength,
+                                     float *error_rate)
+{
+  int ret = 0;
+
+  const char *at_prefix = AT_COMMAND_PREFIX;
+  const char *base_command = AT_COMMAND_CSQ;
+  char command_buffer[AT_COMMAND_BUFFER_STACK_MAX];
+  snprintf(command_buffer,
+           sizeof(command_buffer),
+           "%s%s\r\n",
+           at_prefix,
+           base_command);
+
+  at_serial_port_command_t *at_command =
+    at_serial_port_command_create(command_buffer);
+  if (at_command != NULL) {
+    at_serial_port_execute_command(port, at_command);
+    const char *result = at_serial_port_command_result(at_command);
+
+    snprintf(command_buffer, sizeof(command_buffer), "%s:", base_command);
+    char *parse_ref = strstr(result, command_buffer);
+    if (parse_ref != NULL) {
+      parse_ref += strlen(command_buffer);
+      u32 signal_raw = (u32)strtoul(parse_ref, &parse_ref, 10);
+      *signal_strength = modem_signal_strength_to_dbm(signal_raw);
+      if (*parse_ref++ == ',') {
+        u32 error_raw = (u32)strtoul(parse_ref, NULL, 10);
+        *error_rate = modem_error_rate_to_percent(error_raw);
+      }
+    } else {
+      ret = -1;
```

Comment:
I've yet to see a convincing argument that this style is less error prone... and I definitely find it less obvious to have the handling for an error condition hiding at the end of blob of code, rather than handled immediatey when it's found (from a lexicographical point of view).  Sorry, I'm going probably going to complain about this until we agree on a standard or find other convincing references...